### PR TITLE
Mini-fix of Locate for recursive notations using named variables.

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** bug fix / code reworking

This fixes `Locate "( x , y , .. , z )"` which was otherwise returning nothing. At the same time, we add support for `Locate "( x , y )"` and `Locate "( x , y , z )"` etc., so that it finds `"( x , y , .. , z )"`.

We also reorganize the code for interpreting notations across `notation.ml` and `metasyntax.ml` so as to be able to extend notations string with token information such as `bigint` or `global` in anticipation of #11561.

- [X] Added / updated test-suite
- [ ] Entry added in the changelog

### CODE REVIEW

**Comment by herbelin** [comment]
Note: the fix of `Locate "U"` (in their case `Locate "while"` for `"while"` part of the syntax of an imperative language) would be useful for the next release of Software Foundations. So, it might help them to have it in 8.12.

File: interp/notation.ml [hunk 1]
```diff
@@ -1919,23 +1919,112 @@ let rec raw_analyze_notation_tokens = function
...
+  let toks =
+    if
+      List.exists (function String "_" -> true | _ -> false) toks ||
+      List.for_all (function String id -> Id.is_valid id | _ -> false) toks
+    then
+      (* Only "_ U _" format *)
+      raw_analyze_anonymous_notation_tokens toks
+    else
+      (* Includes the case of only a subset of tokens or an "x 'U' y"-style format *)
+      raw_analyze_notation_tokens toks in
```
Comment (SkySkimmer) [inline interp/notation.ml:1998]:
```suggestion
      raw_analyze_notation_tokens toks 
  in
```

File: interp/notation.ml [hunk 2]
```diff
@@ -1919,23 +1919,112 @@ let rec raw_analyze_notation_tokens = function
...
+    | SProdList (_,l1) :: toks1, SProdList (_, l2) :: toks2 -> aux l1 l2 && aux toks1 toks2
+    | NonTerminal _ :: toks1, SProdList (_,l2) :: toks2 -> aux' toks1 l2 l2 toks2 || aux toks1 toks2
+    | [], [] -> true
+    | (Break _ :: _, _) | (_, Break _ :: _) -> assert false
+    | (Terminal _ | NonTerminal _ | SProdList _) :: _, _ -> false
+    | [], _ -> false
+  and aux' toks1 l2 l2full toks2 = match (toks1, l2) with
+    | Terminal s1 :: toks1, Terminal s2 :: l2 when String.equal s1 s2 -> aux' toks1 l2 l2full toks2
+    | NonTerminal _ :: toks1, [] -> aux' toks1 l2full l2full toks2 || aux toks1 toks2
+    | _ -> false in
```
Comment (SkySkimmer) [inline interp/notation.ml:2017]:
```suggestion
    | _ -> false 
  in
```

**Review by SkySkimmer — COMMENTED** [review]
Seems OK, maybe someone more familiar with notations should have a look though.

File: interp/notation.ml (hunk 2 above)
Comment (herbelin) [reply interp/notation.ml:2018]: It is not the style I usually follow for `let in`, but why not in this case.

**Comment by herbelin** [comment]
> someone more familiar with notations should have a look though.

I guess this means @ejgallego (or @maximedenes if he's back).

File: interp/notation.ml [hunk 3]
```diff
@@ -1777,12 +1777,12 @@ type symbol =
   | Break of int
 
 let rec symbol_eq s1 s2 = match s1, s2 with
-| Terminal s1, Terminal s2 -> String.equal s1 s2
-| NonTerminal id1, NonTerminal id2 -> Id.equal id1 id2
-| SProdList (id1, l1), SProdList (id2, l2) ->
-  Id.equal id1 id2 && List.equal symbol_eq l1 l2
-| Break i1, Break i2 -> Int.equal i1 i2
-| _ -> false
+  | Terminal s1, Terminal s2 -> String.equal s1 s2
```
Comment (ppedrot) [inline interp/notation.ml:1780]: This is not a very standard indentation style in the Coq codebase, is it?

Comment (herbelin) [reply interp/notation.ml:1780]: The file `notation.ml` uses both the
```
let f = function
| ...
```
and
```
let f = function
  | ...
```
indentation style but the second is much more popular (11000 vs 2000 occurrences).

But maybe did I misunderstand your question?

**Comment by maximedenes** [comment]
> I guess this means @ejgallego (or @maximedenes if he's back).

I'm not very familiar with this precise part of the notations system, so I guess it would be better if @ejgallego could have a look. If he's not available, I guess we can still proceed.

**Comment by herbelin** [comment]
Are we exhausting us with reviews? What should we do?

File: interp/notation.ml (hunk 3 above)
Comment (ppedrot) [reply interp/notation.ml:1932]: (This is not a blocking remark in any way, so I am closing this subthread.)

**Review by ejgallego — APPROVED** [review]
LGTM, logic in `interpret_notation_string` could be made a bit more declarative by actually naming the condition in the `if`.

Please rebase to get a clean CI.

**Comment by gares** [comment]
@ejgallego would you mind merging?

**Comment by ejgallego** [comment]
@coqbot: merge now

**Pull request merged**